### ISSUE TITLE
create a run loop so we can use/try the program

### ISSUE THREAD

**Issue by GetContented** [rwobben/paytime#12]
It's probably a good idea to program a run loop that we can offer the user a bunch of options in, and then keep the data in memory while we let them try the functionality out.

I imagine we'll need things like:

- Open an account
- Close an account
- Make a deposit
- Make a withdrawl
- Find out the bank balance
- Print the list of transactions

Did I miss anything?

**Comment by rwobben** [comment]
I do not think so. or it ust be that not one but two events are happing. Like first a deposit and then a withdrawl or two deposits .

**Comment by GetContented** [comment]
If you wanted to do that, wouldn't you just do them in sequence? This runloop would keep its data around for the session until you quit.

**Comment by rwobben** [comment]
Not a problem. We are not working async now , which is not handy on a accounting system

**Comment by GetContented** [comment]
What do you mean?

**Comment by rwobben** [comment]
Whar I mean is for example you can work async. 
This is not handy for a accounting system because then a withdrawl can be handled before a deposit and make a account negative which we do not want. 

So in sequence is better

**Comment by GetContented** [comment]
Sorry, but I don't know what you're saying, sadly.

This *may* be a very important point, though.

In our example... it's not really possible to do these things out of order... the only time we're applying multiple transactions, we're doing so from a list, which has an implicit order baked into the structure.

At the moment, in our "program", we have nothing that can be "executed", so there is no actual runnable program yet, just a bunch of functions. We're just building the components that we'll use when we build our program.

Haskell functions are pure. That means they're just a relation between a set of values and another set of values. Another way to say that is they're parameterized expressions. When you choose some values to plug into the expressions (the functions), you will end up with a result - the result that that function relates for that input value.

This says nothing about time at all. Nothing about sequence. Therefore nothing about async or sync. A Haskell function is a pure mathematical function, just like in maths. In Javascript, Ruby or Clojure, there are things called functions, but they are not actually guaranteed to be pure functions.

One of the properties of a pure function is that its inputs will ALWAYS produce the same outputs. ALWAYS. Absolutely always. In Clojure it's possible to write things they call functions which are not mathematical functions. They're procedures. That is, you can easily write a "fn" that spits out a different value each time you run it. This is literally impossible in Haskell.

Haskell keeps its impure parts totally separate from its pure parts. This is why doing some things that seem easy in other languages seems so difficult in Haskell. It's actually an incredibly useful thing, though, because it forces you to be aware of what you're doing, and it's harder to get into a mess. :)

**Comment by rwobben** [comment]
its good. we mean the same I think.

**Comment by GetContented** [comment]
Ah ok. :) Sorry, I got a bit confused by what you meant :) all good though.

**Comment by rwobben** [comment]
Can we work on this one. and take one part a try.
So first we try to open a account. 

Another thought. We have not write any test now.  I think we need to use test somewhere down the line.

**Comment by GetContented** [comment]
Ok... it's best to go in tiny steps, so first it'd be good to work out how you want it to work?

**Comment by GetContented** [comment]
As to the tests, what tests do you need?

So far the only one that springs to mind is to test if depositing and amount then withdrawing an amount results in an account that has the same balance as before you started.

**Comment by rwobben** [comment]
oke, 

For making a account. I think it's most easy to use the arguments of a function because we did not touch monads at the moment.

So makeAccount "test"  0  schould make a account named test with a balance of zero

**Comment by GetContented** [comment]
Ok, sure, but aren't we talking about how you want the interface to work at the moment?

How should that work?

**Comment by rwobben** [comment]
Correct, at this moment that is all we can test. 

if we make a function that makes a new account. we can test if  a valid account is made and test if a non-valid account is not made.  As far as I understand it makeaccount "test2" schould not work because the balance is missing. 

What do you exactly mean with interface to work at this moment. ? I do not understand the word interface.

**Comment by GetContented** [comment]
Validity is handled by the type system. It's not possible to make a non-valid type.

I mean "user interface" - which is "how the user uses the software".

So... when they run the program, what should it print on the screen? What kind of input should it accept? etc.

**Comment by rwobben** [comment]
When running it schould accept a name and a balance. 
When the validaty passes so the programm is not "chrashing" so no error messages then the user schould see the messsage that account x is made with as balance value y.

**Comment by rwobben** [comment]
so I think something like this : 

   makeAccount name balance 
      account{accountBalance = { AccountName = name , Accountbalance = balance, accountTransaction = []} 
       print $ "The account is made "

**Comment by GetContented** [comment]
Ok... assuming we have just that one feature (create an account), it'd probably be good to show a list of all the accounts made at the beginning of the run loop. That way you can at least see what you're doing...

**Comment by rwobben** [comment]
I have no problem with that. 
Then we need a recursive function to list all accounts and after that some code to add a new account. 
This looks not a very difficult task, 

The only thing I have not figured out is how we can output things like "the account bank is made with as balance value 3000' 
Maybe this way : print $  "the account" ++ bank " is made with as balance value" ++ show (value) ?

**Comment by GetContented** [comment]
yeah, that's sort of right (missing a ++ after bank, and you don't need print if the things are already showable). I'll attach my version and you can compare with yours.

**Comment by GetContented** [comment]
makeAccount :: String -> Amount -> Account
    makeAccount n a = Account {accountName = n, accountAmount = a, accountTransactions = []}

    main :: IO ()
    main = runLoop []
      where runLoop :: [Account] -> IO ()
            runLoop accounts = do
              putStrLn $ "Accounts raw data: " ++ show accounts
              putStrLn "Enter new account name"
              newName <- getLine
              putStrLn "Enter new account balance"
              balStr <- getLine
              let newBal      = read balStr :: Integer
                  newAcct     = makeAccount newName newBal
                  newAccounts = newAcct : accounts
              putStrLn $ "Account " ++ newName ++ " made with balance " ++ (show newBal) ++ "."
              runLoop newAccounts

**Comment by GetContented** [comment]
Notice how we're evaluating runLoop newAccounts at the end? That feeds the same list back into the function again to create a new account (to quit the program, use CTRL-C until we implement a quit feature).

Notice we can put a type signature into a where clause (which we're doing for the definition of the runLoop helper function).

Each line in this IO monad must result in an IO type, unless it uses <- or let. <- can pull and set values out of the IO (which we're doing with getLine), and let just defines variables which can be used after they're defined.

At the end, we call our runLoop again, but with the new accounts list that has our newly added account in it (created in the line newAccounts = newAcct : accounts  which as you know uses (:) to prepend an element to a list).

**Comment by rwobben** [comment]
Thanks, 

That part works now with a few adjustments. Some names were wrong. 

So now the next part , delete a account 

So I think we need now some sort of menu 
First choose for : 

1  Add Account 
2  Delete Account 

On add Account the questions for adding needs to be displayed. 
on delete Account the to be deleted account need to be asked and and check if the account does exists. If so, delete it , if not. give the user a message 
On both the first menu needs to be displayed. 

So I think that the questions need to be in a seperate function  otherwise it will be a very big main function.

**Comment by rwobben** [comment]
Second thing. 
I tried a wierd input : 

Accounts raw data: []
Enter new account name
gek
Enter new account balance
a
paytime-exe: Prelude.read: no parse

But can be make a better error message here,
A user cannot understand this

**Comment by rwobben** [comment]
I was thinking. If we do al things with only ghci , I think the code will be very complex.
Maybe a idea to look for a easy web framework and make this work there.
Then we can make for every challenge a seperate page.

**Comment by GetContented** [comment]
I don't think the code will be very complex.

Yeah, read is the simplest "parser", but it's not able to cope with things it can't parse as the expected type. It's good to ask why we're doing this, though... this is just so we can try out our functions in a context, isn't it? In other words, we're not making it for end users, we're making it for ourselves, because it'll make you feel like you're making a real program and help you to understand how things would work in a real program.

If you *must* have validation of this kind, we can use the readMaybe function from the Text.Read library/module. To use this, you need to know how the Maybe type works, because it returns a "Maybe a" value.

Like you say, it's not the "end game" so to speak, so it probably isn't worth spending much time on - just a  quick and dirty thing so we can try out our functions.

Scotty is about the simplest web app, and it's already many many times more complex than anything you can do yet, so this might defeat the entire purpose of doing this, which is to help you to learn Haskell, and use this application as a set of real world-like examples so you can learn in context.

**Comment by rwobben** [comment]
oke, 

I was thinking about Servant with blaze-html. 

Not complex. Multiple choices and the functions that have to do it. 
I have not a problem with quick but maybe with dirty. 
I want to learn haskell well 

Or do you want to change main for every function to look if it's working.

**Comment by GetContented** [comment]
Writing code in Servant is not easy for a beginner. In fact, writing code in *any* of the web frameworks is not easy for a beginner.

"Quick and dirty" just means we're going to throw it away eventually, so why bother polishing it? Building ourselves a little test application is just a prototyping exercise. You've said you want to put it into a web application. To do that requires understanding a lot more Haskell, though.

If you want a step-by-step example of building an extremely simple web application (it just shows a list of blog entries from static data), check this out: http://www.genericoverlords.com/snap_notes

**Comment by rwobben** [comment]
oke, 

Can you show me then why the code will be not complex with adding and deleting in one main function. 

A web framework was just a idea to make things easler but apperently im was wrong.

**Comment by GetContented** [comment]
Ok... well this is one way of building it into something that has a menu... it's only like 50 lines of code. I don't think that's much code. You can see adding each new "feature" is probably only going to add another 14 or so lines at the most... 

    tryAgainErrorMessage :: String
    tryAgainErrorMessage = "Incorrect menu item. Try again."

    main :: IO ()
    main = runLoop []
      where runLoop :: [Account] -> IO ()
            runLoop accounts = do
              let canDelete = length accounts > 0
              putStrLn "Menu."
              putStrLn $ "CA: Create Account." ++
                (if canDelete then "\nDA: Delete Account." else "")
                ++ "\nSA: Show Accounts. \nQ: Quit"
              menuChoice <- getLine
              case menuChoice of
                "CA" -> do
                          newAccounts <- doCreateAccounts accounts
                          runLoop newAccounts
                "DA" -> if canDelete
                        then do
                              newAccounts <- doDeleteAccounts accounts
                              runLoop newAccounts
                        else do
                              putStrLn tryAgainErrorMessage
                              runLoop accounts
                "SA" -> do
                          putStrLn $ "Accounts raw data: " ++ show accounts
                          runLoop accounts
                "Q"  -> return ()
                _    -> do
                          putStrLn tryAgainErrorMessage
                          runLoop accounts

    doCreateAccounts :: [Account] -> IO [Account]
    doCreateAccounts accounts = do
      putStrLn "Enter new account name"
      newName <- getLine
      putStrLn "Enter new account balance"
      balStr <- getLine
      let newBal      = read balStr :: Integer
          newAcct     = makeAccount newName newBal
          newAccounts = newAcct : accounts
      putStrLn $ "Account " ++ newName ++ " made with balance " ++ (show newBal) ++ "."
      return newAccounts

    doDeleteAccounts :: [Account] -> IO [Account]
    doDeleteAccounts accounts = do
      putStrLn "Nothing here yet..."
      return []

**Comment by GetContented** [comment]
I should mention for the above to work, we have add a deriving (Show) to a couple of the data types. This will make them instances of Show automatically which means we can convert them to String types by using the show function on them.

**Comment by rwobben** [comment]
I found out already by the error messages on the add function.
Now I have to think how I can check if a account exist. 
I thought  deleteAccount `elem` accounts would work but not. I get a type error.

**Comment by GetContented** [comment]
I don't understand your thought process here... you're better off putting "delete accounts" as a separate issue, aren't you?

Then there can be two parts to that issue... one is writing the pure function that will delete an accounts at a particular index, or by name... the other is extending the runLoop's doDeleteAccounts function so that it works by calling the deleteAccount function.

**Comment by GetContented** [comment]
I think this is pretty much finished, isn't it? Are we missing anything from this list? We can close this now, right?

**Comment by rwobben** [comment]
Nope, We have to make a list of transactions. 
That is what I working on to make a nice layout to print all the transactions of a account.
